SYNC-MUTATION.md

```diff
@@ -0,0 +1,335 @@
+
+For example, given a snippet of code:
+
+```js
+function *gen() {
+  yield computeResult();
+  yield computeResult2();
+}
+```
+
+If we want to scope the `computeResult` and `computeResult2` calls with a new
+AsyncContext value, it needs non-trivial refactor:
+
+```js
+const asyncVar = new AsyncContext.Context();
+
+function *gen() {
+  const span = createSpan();
+  yield asyncVar.run(span, () => computeResult());
+  yield asyncVar.run(span, () => computeResult2());
+  // ...or
+  yield* asyncVar.run(span, function *() {
+    yield computeResult();
+    yield computeResult2();
+  });
+}
+```
+
+`.run(val, fn)` creates a new function body. The new function environment
+is not equivalent to the outer environment and can not trivially share code
+fragments between them. Additionally, `break`/`continue`/`return` can not be
+refactored naively.
+
+It will be more intuitive to be able to insert a new line and without refactor
+existing code snippet.
+
+```js
+const asyncVar = new AsyncContext.Context();
+
+function *gen() {
+  asyncVar.set(createSpan(i));
+  yield computeResult(i);
+  yield computeResult2(i);
+}
+```
+
+## The set semantics
+
+With the name of `set`, this method actually doesn't modify existing async
+context snapshots, similar to consecutive `run` operations. For example, in
+the following case, `set` doesn't change the context variables in async tasks
+created just prior to the mutation:
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("main");
+new AsyncContext.Snapshot() // snapshot 0
+console.log(asyncVar.get()); // => "main"
+
+asyncVar.set("value-1");
+new AsyncContext.Snapshot() // snapshot 1
+Promise.resolve()
+  .then(() => { // continuation 1
+    console.log(asyncVar.get()); // => 'value-1'
+  })
+
+asyncVar.set("value-2");
+new AsyncContext.Snapshot() // snapshot 2
+Promise.resolve()
+  .then(() => { // continuation 2
+    console.log(asyncVar.get()); // => 'value-2'
+  })
+```
+
+The value mapping is equivalent to:
+
+```
+⌌-----------⌍ snapshot 0
+|   'main'  |
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 1
+| 'value-1' |  <---- the continuation 1
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 2
+| 'value-2' |  <---- the continuation 2
+⌎-----------⌏
+```
+
+This trait is important with both `run` and `set` because mutations to
+`AsyncContext.Variable`s must not mutate prior `AsyncContext.Snapshot`s.
+
+> Note: this also applies to [`ContinuationVariable`][]
+
+### Decouple mutation with scopes
+
+To preserve the strong scope guarantees provided by `run`, an additional
+constraint can also be put to `set` to declare explicit scopes of mutation.
+
+A dedicated `AsyncContext.contextScope` can be decoupled with `run` to open a
+mutable scope with a series of `set` operations.
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
```

Comment:
My expectation would be that top-level would have an implicit scope so there should never be a point where `store.set(...)` can throw. This is also a requirement for solving the top-level configuration sharing problem Matteo described.

Rather, I would want for this `AsyncContext.contextScope(...)` thing to be an option for a user to provide _additional_ scopes when they want ones more narrow than the runtime already provides. The runtime-provided scopes I'm thinking of would be module/script top-level and async functions. I believe everything else could just inherit the scope as it inherits context.